utils/redis.ts

```diff
@@ -0,0 +1,96 @@
+/* eslint-disable no-invalid-this */
+import { identity, pipe } from "fp-ts/lib/function";
+import * as O from "fp-ts/lib/Option";
+import * as redis from "redis";
+import { IConfig } from "./config";
+
+const DEFAULT_REDIS_PORT = "6379";
+
+export type RedisClient = redis.RedisClientType | redis.RedisClusterType;
+
+export class RedisClientFactory {
+  protected readonly config: IConfig;
+  // eslint-disable-next-line functional/prefer-readonly-type
+  protected redisClient: RedisClient | undefined;
+
+  constructor(config: IConfig) {
+    this.config = config;
+  }
+
+  public readonly getInstance = async (): Promise<RedisClient> => {
+    if (!this.redisClient) {
+      // eslint-disable-next-line functional/immutable-data
+      this.redisClient = await pipe(
+        this.config.isProduction,
+        O.fromPredicate(identity),
+        O.chainNullableK(_ => this.config.REDIS_CLUSTER_ENABLED),
+        O.chain(O.fromPredicate(identity)),
+        O.map(() =>
+          this.createClusterRedisClient(
+            this.config.REDIS_URL,
+            this.config.REDIS_PASSWORD,
+            this.config.REDIS_PORT
+          )
+        ),
+        O.getOrElse(() =>
+          this.createSimpleRedisClient(
+            this.config.REDIS_URL,
+            this.config.REDIS_PASSWORD,
+            this.config.REDIS_PORT,
+            this.config.REDIS_TLS_ENABLED
+          )
+        )
+      );
+    }
+    return this.redisClient;
+  };
+
+  protected readonly createSimpleRedisClient = async (
+    redisUrl: string,
+    password?: string,
+    port?: string,
```

Comment:
why it's a `string` ? you can accept here directly a `number` (and set here the default port)

Reply:
Usually env vars come as Strings and are parsed in the config as NumberFromString.
Here the code, as said above, is ported "as is" from another repo. And we keep porting as it is to have an easy search about it.
Lets track activity and make a common package and clean it as we prefer.